src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -520,17 +567,66 @@ private fun RsCallExpr.expectedParamsCount(): Int? {
     }
 }
 
+private fun RsMethodCallExpr.expectedMutableSelfParameter(): Boolean {
+    val fn = reference.resolve() as? RsFunction ?: return false
+    val self = fn.selfParameter ?: return false
+    return self.isMut && self.isRef
+}
+
 private fun RsMethodCallExpr.expectedParamsCount(): Int? {
     val fn = reference.resolve() as? RsFunction ?: return null
     if (fn.queryAttributes.hasCfgAttr()) return null
     return if (fn.isInherentImpl) fn.valueParameterList?.valueParameterList?.size else null
 }
 
+private val RsCallExpr.declaration: RsFunction?
+    get() = (expr as? RsPathExpr)?.path?.reference?.resolve() as? RsFunction
+
+private fun RsExpr.declaration(): RsCompositeElement? {
+    if (this is RsPathExpr) {
+        return this.path.reference.resolve()
+    }
+    if (this is RsFieldExpr) {
+        return this.expr.declaration()
+    }
+    return null
+}
```

Comment:
I'm not sure, but shouldn't these 2 functions reside in `ResolveEngine`? What's your opinion @matklad ?